File: app/src/main/java/com/botijasoftware/utils/Sprite.java
package com.botijasoftware.utils;

public class Sprite {

	protected Rectanglef mRectangle;
	protected Texture mTexture;
	protected boolean mFlip;
	protected Vector2 mScale = new Vector2(1.0f, 1.0f);
	protected float mRotation = 0.0f;
	protected ColorRGBAb mColor;
	protected boolean mVisible = true;


	public Sprite(Rectanglef rectangle, Texture texture, float scalex, float scaley, float rotation, ColorRGBAb color, boolean flip) {
	
		mRotation = rotation;
		mFlip = flip;

		mRectangle = rectangle;
		setTexture(texture);
		setRectangle(rectangle);
        setScale(scalex, scaley);
		setColor( color );

	}

    public Sprite(Rectanglef rectangle, Texture texture, float scale, float rotation, ColorRGBAb color, boolean flip) {
        this(rectangle, texture, scale, scale, rotation, color, flip);
    }
	
	public Sprite(Rectanglef rectangle, Texture texture) {
		this(rectangle, texture, 1.0f, 0.0f, new ColorRGBAb((byte) 255, (byte) 255, (byte) 255, (byte) 255), false);
	}
	
	public Sprite(Rectanglef rectangle, Texture texture, boolean flip) {
		this(rectangle,texture, 1.0f, 0.0f, new ColorRGBAb((byte)255,(byte)255,(byte)255,(byte)255), flip);
	}
	
	public Sprite(Rectanglef rectangle, Texture texture, ColorRGBAb color) {
		this(rectangle,texture, 1.0f, 0.0f, color, false);
	}
	
	public Sprite(Rectanglef rectangle, Texture texture, float scale) {
		this(rectangle,texture, scale, 0.0f, new ColorRGBAb((byte)255,(byte)255,(byte)255,(byte)255), false);
	}
	
	public Sprite(Rectanglef rectangle, Texture texture, float scale, float rotation) {
		this(rectangle,texture, scale, rotation, new ColorRGBAb((byte)255,(byte)255,(byte)255,(byte)255), false);
	}
	
	public Sprite(Rectanglef rectangle, Texture texture, float scale, float rotation, boolean flip) {
		this(rectangle,texture, scale, rotation, new ColorRGBAb((byte)255,(byte)255,(byte)255,(byte)255), flip);
	}
	
	
	public void setTexture(Texture texture) {
		mTexture = texture;
				
		TextureCoords coords = mTexture.getTextureCoords();
		
		if (mFlip) {
			coords.flipVertical();
		}

	}
	
	public void setTexture(Texture texture, boolean flip) {
		
		mFlip = flip;
		setTexture(texture);	
	}
	
	public void setRectangle(Rectanglef rectangle) {
	
		mRectangle = rectangle;
		
		float w = mRectangle.width * mScale.X;
		float h = mRectangle.height * mScale.Y;

	}
	
	public void setPosition(float x, float y) {
		mRectangle.X = x;
		mRectangle.Y = y;
	}
	
	public void setPosition(Vector2 v) {
		mRectangle.X = v.X;
		mRectangle.Y = v.Y;
	}
	
	public void setCenter(Vector2 v) {
		mRectangle.setCenter(v);
	}
	
	public void setCenter(float x, float y) {
		mRectangle.setCenter(x, y);
	}
	
	public void setSize(float nw, float nh) {
		mRectangle.width = nw;
		mRectangle.height = nh;
		
		setRectangle(mRectangle);
	}
	
	public void setSize(Vector2 v) {
		mRectangle.width = v.X;
		mRectangle.height = v.Y;

		setRectangle(mRectangle);
	}
	
	public void setScale(float scalex, float scaley) {
		mScale.X = scalex;
        mScale.Y = scaley;
		setRectangle(mRectangle);
	}

    public void setScale(Vector2 scale) {
        mScale.X = scale.X;
        mScale.Y = scale.Y;
        setRectangle(mRectangle);
    }
	
	public void setRotation(float rotation) {
		mRotation = rotation;
	}
	
	public void setColor(ColorRGBAb color) {
		
		mColor = color;

	}
	
	public void setAlpha(float alpha) {
		mColor.A = (byte) (255*alpha);
		setColor(mColor);
	}
	
	public void setAlpha(byte alpha) {
		mColor.A = alpha;
		setColor(mColor);
	}

	
	public Rectanglef getRectangle() {
		return mRectangle;
	}
	
	public void show() {
		mVisible = true;
	}
	
	public void hide() {
		mVisible = false;
	}
	
	public boolean isVisible() {
		return mVisible;
	}

    public float getFloatAlpha() {
        return (mColor.A / 255.0f);
    }

    public byte getByteAlpha() {
        return mColor.A;
    }

	public Texture getTexture() {
		return mTexture;
	}


	public void move( float x, float y) {
		mRectangle.X += x;
		mRectangle.Y += y;
	}


	public void scale( float x, float y) {
		mScale.X *= x;
		mScale.Y *= y;
	}


}

File: app/src/main/java/com/botijasoftware/utils/Sprite.kt
package com.botijasoftware.utils

open class Sprite(var mRectangle: Rectanglef, texture: Texture, var scalex: Float, var scaley: Float, rotation: Float, color: ColorRGBAb, var mFlip: Boolean) {
    var mTexture: Texture
    var mScale = Vector2(1.0f, 1.0f)
    var mAnchor = Vector2(0.5f, 0.5f)
    var mRotation = 0.0f
    lateinit var mColor: ColorRGBAb
    var isVisible = true
        protected set


    init {

        mRotation = rotation
        mTexture = texture
        //rectangle = mRectangle
        setScale(scalex, scaley)
        setColor(color)

    }

    @JvmOverloads constructor(rectangle: Rectanglef, texture: Texture, scale: Float, rotation: Float = 0.0f, color: ColorRGBAb = ColorRGBAb(255.toByte(), 255.toByte(), 255.toByte(), 255.toByte()), flip: Boolean = false) : this(rectangle, texture, scale, scale, rotation, color, flip) {}

    constructor(rectangle: Rectanglef, texture: Texture) : this(rectangle, texture, 1.0f, 0.0f, ColorRGBAb(255.toByte(), 255.toByte(), 255.toByte(), 255.toByte()), false) {}

    constructor(rectangle: Rectanglef, texture: Texture, flip: Boolean) : this(rectangle, texture, 1.0f, 0.0f, ColorRGBAb(255.toByte(), 255.toByte(), 255.toByte(), 255.toByte()), flip) {}

    constructor(rectangle: Rectanglef, texture: Texture, color: ColorRGBAb) : this(rectangle, texture, 1.0f, 0.0f, color, false) {}

    constructor(rectangle: Rectanglef, texture: Texture, scale: Float, rotation: Float, flip: Boolean) : this(rectangle, texture, scale, rotation, ColorRGBAb(255.toByte(), 255.toByte(), 255.toByte(), 255.toByte()), flip) {}

    fun setTexture(texture: Texture, flip: Boolean) {

        mFlip = flip
        this.texture = texture
    }

    fun setPosition(x: Float, y: Float) {
        mRectangle.X = x
        mRectangle.Y = y
    }

    fun setPosition(v: Vector2) {
        mRectangle.X = v.X
        mRectangle.Y = v.Y
    }

    fun setCenter(v: Vector2) {
        mRectangle.center = v
    }

    fun setCenter(x: Float, y: Float) {
        mRectangle.setCenter(x, y)
    }

    open fun setSize(nw: Float, nh: Float) {
        mRectangle.width = nw
        mRectangle.height = nh

        rectangle = mRectangle
    }

    open fun setSize(v: Vector2) {
        mRectangle.width = v.X
        mRectangle.height = v.Y

        rectangle = mRectangle
    }

    fun setScale(scalex: Float, scaley: Float) {
        mScale.X = scalex
        mScale.Y = scaley
        rectangle = mRectangle
    }

    fun setScale(scale: Vector2) {
        mScale.X = scale.X
        mScale.Y = scale.Y
        rectangle = mRectangle
    }

    fun setRotation(rotation: Float) {
        mRotation = rotation
    }

    fun setColor(color: ColorRGBAb) {

        mColor = color

    }

    fun setAlpha(alpha: Float) {
        mColor.A = (255 * alpha).toByte()
        setColor(mColor)
    }

    fun setAlpha(alpha: Byte) {
        mColor.A = alpha
        setColor(mColor)
    }


    open var rectangle: Rectanglef
        get() = mRectangle
        set(rectangle) {

            mRectangle.X = rectangle.X
            mRectangle.Y = rectangle.Y
            mRectangle.width = rectangle.width * mScale.X
            mRectangle.height = rectangle.height * mScale.Y

        }

    fun show() {
        isVisible = true
    }

    fun hide() {
        isVisible = false
    }

    val floatAlpha: Float
        get() = mColor.A / 255.0f

    val byteAlpha: Byte
        get() = mColor.A

    open var texture: Texture
        get() = mTexture
        set(texture) {
            mTexture = texture

            val coords = mTexture.textureCoords

            if (mFlip) {
                coords.flipVertical()
            }

        }


    fun move(x: Float, y: Float) {
        mRectangle.X += x
        mRectangle.Y += y
    }


    fun scale(x: Float, y: Float) {
        mScale.X *= x
        mScale.Y *= y
    }

    fun setAnchor(x: Float, y: Float) {
        mAnchor.X = x
        mAnchor.Y = y
    }

    var anchor: Vector2
        get() = mAnchor
        set(anchor) {
            mAnchor.X = anchor.X
            mAnchor.Y = anchor.Y
        }

}
